ir/visitor.cpp

```diff
@@ -19,54 +19,69 @@ limitations under the License.
 #include "lib/log.h"
 
 class Visitor::ChangeTracker {
-    // FIXME -- this code is really incomprehensible due to all the pairs/first/second stuff
-    // unfortunatelly maps use pairs all over the place, which is where they come from.
-    typedef unordered_map<const IR::Node *, std::pair<bool, const IR::Node *>>  visited_t;
+    struct visit_info_t {
+        bool visit_in_progress;
+        const IR::Node *result;
+    };
+    typedef unordered_map<const IR::Node *, visit_info_t>  visited_t;
     visited_t           visited;
 
  public:
-    struct change_t {
-        bool                                    valid;
-        std::pair<visited_t::iterator, bool>    state;  // result of visited.emplace
-        change_t() : valid(false) {}
-        change_t(visited_t *visited, const IR::Node *n) : valid(true),
-            state(visited->emplace(n, std::make_pair(false, n))) {
-                if (!state.second && !state.first->second.first)
-                    BUG("IR loop detected "); }
-        explicit operator bool() { return valid; }
-        bool done() { return !state.second; }
-        const IR::Node *orig() { return state.first->first; }
-        const IR::Node *result() { return state.first->second.second; }
-    };
-    bool done(const IR::Node *n) const { return visited.count(n) && visited.at(n).first; }
-    const IR::Node *result(IR::Node *n) const { return visited.at(n).second; }
-    change_t track(const IR::Node *n) { return change_t(&visited, n); }
-    void start(change_t &change) { change.state.first->second.first = false; }
-    bool finish(change_t &change, const IR::Node *orig, const IR::Node *final) {
-        if (!change.valid || (change.state.first = visited.find(orig)) == visited.end())
+
+    /* Begin tracking `n` during a visiting pass.  Use `finish` to mark `n` as
```

Comment:
The more things in Doxygen, the better. (Plus, it's good to just get in the habit of writing Oxygen comments.)